### PR TITLE
Add transformation guide

### PR DESCRIPTION
**Pull request by gjtorikian**
This is a guide that builds off of https://github.com/atomist/docs/pull/287, and discusses code transformations.

It's very nearly complete, but there's one thing I couldn't figure out: how to access parameters within the generator function. I would like the users' provided repo name to be the replacing text. I see the default arguments showing up in the command:

```
atomist generate-node-express-project

Node Express Project

Options:
  --help, -h, -?        Show help                                                          [boolean]
  --target-repo
  --target-description
  --target-visibility
  --target-apiUrl
  --target-owner
```

But I don't know how to get at them! @jessitron, would you be able to help out by providing a small code example for that here? :pray:

### CODE REVIEW

File: docs/developer/first-transform.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+A code transform allows you to change any of the files in a seed repository at the time of a project's generation. You can use it to do anything from changing the copyright year in source code header licenses to manipulating the AST of the files entirely.
+
+In this example, we'll apply a code transform that does two things to a Node project: first, it'll update the README.md to note the project's name, and second, it'll also update the package.json file to use the project's name, too.
+
+## Prerequisites
+
+You should already have [a blank SDM](/developer/sdm/#creating-an-sdm-project), though you can also use an existing one if you'd like.
+
+We'll be using the [atomist-seeds/express-es6-rest-api](https://github.com/atomist-seeds/express-es6-rest-api) repository as our seed. Please don't use this seed for your own Express projects, as this is just a demo app (and likely full of outdated packages!).
```
Comment (jessitron) [inline docs/developer/first-transform.md]: is that a problem? can we keep this seed up?

File: docs/developer/first-transform.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+## Registering a generator
+
+Since we've already gone through this before, we'll simply provide the generic generator boilerplate to get started.  Open up your SDM, and add a new file called `lib/node_generator.ts`. Paste the following lines to get started with your generator:
```
Comment (jessitron) [inline docs/developer/first-transform.md]: is it boilerplate? this is a sample generator definition, I don't think it's a bunch of boilerplate?

File: docs/developer/first-transform.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  name: "Node Express Project",
+  intent: "generate-node-express-project",
+  startingPoint: GitHubRepoRef.from({
+      owner: "atomist-seeds",
+      repo: "express-es6-rest-api",
+  }),
+  transform: []
+};
+```
+
```
Comment (jessitron) [inline docs/developer/first-transform.md]: should we try the generator first? i really want to try it first. (we have to register it with the SDM)

**Comment by jessitron** [comment]
ok good news and bad news. 
The good news is, I got this working

The bad news is, it looks like this:

```typescript
    transform: [
        async (project, papi) =>
            updateTitle("README.md", papi.parameters.target.repoRef.repo)(project, papi),
    ],
```

File: docs/developer/first-transform.md (hunk 1 above)
Comment (gjtorikian) [reply docs/developer/first-transform.md]: I mean, we could drop that parenthetical, but I think it's unrealistic to expect that the seed will be maintained by Atomist, for example, a node package has a CVE. The seed is just there for an example and I don't want to give the impression that it should be relied upon for a production app.

**Comment by gjtorikian** [comment]
What does `papi` stand for?

**Comment by ipcrm** [comment]
@gjtorikian https://atomist.github.io/sdm/interfaces/_lib_api_registration_pushawareparametersinvocation_.pushawareparametersinvocation.html

**Comment by jessitron** [comment]
OK, I just hit the release button on the markdown pack.

This is a better impl:

```typescript

import { updatePageTitle } from "@atomist/sdm-pack-markdown";
...

    transform: [
        async (project, papi) => {
            await updatePageTitle(project, "README.md", papi.parameters.target.repoRef.repo);
        },
    ],
```

What i like about this is, it defines the transform as a function which does a named thing.
This lets people do other things with the project, it's a place they can build from, which `updateTitle` was not, it could only do that one thing and didn't lead into "now make it do something else of your choice"

they'll still need introduced to the Project API, there's a page for that iirc

**Comment by gjtorikian** [comment]
Awesome, thank you for that, @jessitron. When you've got a chance I think this is ready for another review.

**Comment by jrday** [comment]
@jessitron is this ready to be merged or more review needed?

**Comment by jessitron** [comment]
@jrday I say merge it

**Pull request merged**